framework/src/tinyalsa/tinyalsa.c

```diff
@@ -1109,53 +1188,134 @@ int pcm_stop(struct pcm *pcm)
 		return -EINVAL;
 	}
 #ifdef CONFIG_AUDIO_MULTI_SESSION
-	if (ioctl(pcm->fd, AUDIOIOC_STOP, (unsigned long)pcm->session) < 0)
+	if (!pcm->running && ioctl(pcm->fd, AUDIOIOC_STOP, (unsigned long)pcm->session) < 0)
 #else
-	if (ioctl(pcm->fd, AUDIOIOC_STOP, 0) < 0)
+	if (!pcm->running && ioctl(pcm->fd, AUDIOIOC_STOP, 0) < 0)
 #endif
 		return oops(pcm, errno, "cannot stop channel");
 
+	/* Remove any pending messages from the message queue */
+	struct timespec st_time;
+	struct audio_msg_s msg;
+	unsigned int size;
+	int prio;
+	while (1) {
+		clock_gettime(CLOCK_REALTIME, &st_time);
+		size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
```

Comment:
Why current time is passed as wait time for mq_timedreceive?

Reply:
We want to remove any messages already on the queue. This is because user can make calls like start() -> .... -> stop() -> start(). So, we want the queue to be clean between subsequent calls to start() / stop(). Since we only want to remove any messages already on the queue, we are using current time. We dont want to wait for any new messages.

Reply:
Can we get messages from previous start after sometime? For example, the buffer was still in queue to be processed when pcm_stop was called.

Reply:
Once we call AUDIOIOC_STOP, we assume that the driver wont process any more buffers. 
The codec hw will be reset by the driver and any pending or in-process transfers will be dropped. 
So, the codec driver will not put any new messages on the queue. 

The above code only checks if there were already any messages on the queue and removes them. We are using the loop because tinyara mqueue does not provide a utility API to empty the queue.